src/citrine/builders/design_spaces.py

```diff
@@ -0,0 +1,205 @@
+            raise ValueError('values in {} are not between 0 and 1'.format(kk))
+
+    non_balance_comps = [comp for comp in list(formulation_grids.keys())
+                         if comp != balance_component]
+    non_balance_descriptors = [dd for dd in descriptors
+                               if dd.key != balance_component]
+    non_balance_grids = {
+        k: v for k, v in formulation_grids.items()
+        if k in non_balance_comps
+    }
+
+    # Start by making a naive product design space of non-balance components
+    form_ds = pd.DataFrame(
+        enumerate_cartesian_product(
+            design_grids=non_balance_grids,
+            descriptors=non_balance_descriptors
+        ).data
+    )
+
+    # Re-calculate the balance component column as 1-all others
+    form_ds[balance_component] = form_ds.apply(
+        lambda r: 1 - sum([r[col] for col in non_balance_comps]),
+        axis=1
+    )
+
+    # Order columns so balance component is first
+    form_ds = form_ds[[balance_component] + non_balance_comps]
+
+    # Eliminate out-of-range rows
+    balance_range = (
+        min(formulation_grids[balance_component]),
+        max(formulation_grids[balance_component])
+    )
+    form_ds = form_ds[form_ds[balance_component] >= balance_range[0]]
+    form_ds = form_ds[form_ds[balance_component] <= balance_range[1]]
+    data = form_ds.to_dict('records')
+    design_space = EnumeratedDesignSpace(name=name,
+                                         description=description,
+                                         descriptors=descriptors,
+                                         data=data)
+    return design_space
+
+
+def cartesian_join_design_spaces(
+    design_space_list: List[EnumeratedDesignSpace],
+    name: str = 'Joined enumerated design space',
+    description: str = '',
+) -> EnumeratedDesignSpace:
+    """[ALPHA] Cartesian join of multiple enumerated design spaces.
+
+    Given a list of multiple input EnumeratedDesignSpaces, create a new
+    EnumeratedDesignSpace that is the Cartesian product of candidates from each
+    input design space.
+
+    Ultimate length will be the product of the number of candidates in each
+    individual design space
+
+    Parameters
+    ----------
+    design_space_list: List[EnumeratedDesignSpace]
+        A list of EnumeratedDesignSpaces
+
+    """
+    # Test for duplicate or invalid descriptor keys
+    all_keys = []
+    for ds in design_space_list:
+        if 'join_key' in ds.data[0].keys():
+            raise ValueError('"join_key" cannot be a descriptor key when using this function')
+        all_keys.extend(list(ds.data[0].keys()))
+    if len(all_keys) != len(set(all_keys)):
+        raise ValueError('Duplicate keys are not allowed across design spaces')
+
+    # Convert data fields of EDS into DataFrames to prep for join
+    data_list = [ds.data for ds in design_space_list]
+    ds_list = [pd.DataFrame(data) for data in data_list]
```

Comment:
I think these lines could be merged.

Reply:
oh wow line 181 was quite possibly the laziest thing I have ever done haha